File: M3Tools/Components/Database/ListenerDatabase.vb
﻿Imports System.ComponentModel

Namespace Database
	' TODO: Revamp this area as well
	Public NotInheritable Class ListenerDatabase
		Private Const path As String = "listeners"

		<EditorBrowsable()>
		<Description("The username to use for the database connection")>
		<SettingsBindable(True)>
		Public Property Username As String
			Get
				Return dbConnection.Username
			End Get
			Set(value As String)
				dbConnection.Username = value
			End Set
		End Property

		'The password to use for the database connection
		<PasswordPropertyText(True)>
		<SettingsBindable(True)>
		<Description("The password to use for the database connection")>
		Public Property Password As String
			Get
				Return dbConnection.Password
			End Get
			Set(value As String)
				dbConnection.Password = value
			End Set
		End Property

		<Bindable(True)>
		<Description("The initial catalog to use for the database connection")>
		<SettingsBindable(True)>
		Public Property BaseUrl As String
			Get
				Return dbConnection.BaseUrl
			End Get
			Set(value As String)
				dbConnection.BaseUrl = value
			End Set
		End Property

		Public Sub AddListener(firstName As String, lastName As String, email As String)
			AddListener(New Types.Listener(-1, firstName, lastName, email))
		End Sub

		Public Sub AddListener(name As String, email As String)
			AddListener(New Types.Listener(-1, name, email))
		End Sub

		Public Sub AddListener(listener As Types.Listener)
			dbConnection.Consume(Method.Post, $"/{path}", JSON.ConvertToJSON(listener))
		End Sub

		Public Sub RemoveListener(listenerID As Integer)
			If Not Utils.ValidID(listenerID) Then
				Throw New ArgumentException($"ID values must be greater than or equal to {My.Settings.MinID}")
			End If

			dbConnection.Consume(Method.Delete, $"/{path}/{listenerID}")
		End Sub

		Public Sub UpdateListener(listenerID As Integer, fName As String, lName As String, email As String)
			UpdateListener(listenerID, $"{fName} {lName}", email)
		End Sub

		Public Sub UpdateListener(listenerID As Integer, name As String, email As String)
			If Not Utils.ValidID(listenerID) Then
				Throw New ArgumentException($"ID values must be greater than or equal to {My.Settings.MinID}")
			End If

			UpdateListener(New Types.Listener(listenerID, name, email))
		End Sub

		Public Sub UpdateListener(listener As Types.Listener)
			Throw New NotImplementedException("UpdateListener")
		End Sub

		Public Function GetListeners() As Types.DBEntryCollection(Of Types.Listener)
			Return dbConnection.Consume(Of Types.DBEntryCollection(Of Types.Listener))(Method.Get, $"/{path}").Result
		End Function

		Public Function GetListener(listenerID As Integer) As Types.Listener
			If Not Utils.ValidID(listenerID) Then
				Throw New ArgumentException($"ID values must be greater than or equal to {My.Settings.MinID}")
			End If

			Return dbConnection.Consume(Of Types.Listener)(Method.Get, $"/{path}/{listenerID}").Result
		End Function
	End Class
End Namespace

File: M3Tools/Components/Database/ListenerDatabase.cs
using System;
using System.Threading.Tasks;

using SPPBC.M3Tools.M3API;

namespace SPPBC.M3Tools.Database
{
	/// <summary>
	/// 
	/// </summary>
	public partial class ListenerDatabase
	{
		public ListenerDatabase() : base() => InitializeComponent();

		/// <summary>
		/// Retrieves a listener based on the provided listener ID
		/// </summary>
		/// <param name="listenerID"></param>
		/// <param name="ct"></param>
		/// <returns></returns>
		/// <exception cref="ArgumentException"></exception>
		public async Task<Types.Listener> GetListener(int listenerID, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(listenerID)
				? await ExecuteWithResultAsync<Types.Listener>(System.Net.Http.HttpMethod.Get, string.Join(Paths.Separator, Paths.Listeners, listenerID), string.Empty, ct)
				: throw new ArgumentException($"Invalid ListenerID provided");

		/// <summary>
		/// Retrieves the complete list of listeners
		/// </summary>
		/// <returns></returns>
		public async Task<Types.ListenerCollection> GetListeners(System.Threading.CancellationToken ct = default)
			=> await ExecuteWithResultAsync<Types.ListenerCollection>(System.Net.Http.HttpMethod.Get, Paths.Listeners, string.Empty, ct);

		/// <summary>
		/// Add a new listener to the database
		/// </summary>
		/// <param name="listener"></param>
		/// <param name="ct"></param>
		public async Task<bool> AddListener(Types.Listener listener, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Post, Paths.Listeners, JSON.ConvertToJSON(listener), ct);

		/// <summary>
		/// Remove a listener from the database
		/// </summary>
		/// <param name="listenerID"></param>
		/// <param name="ct"></param>
		/// <exception cref="ArgumentException"></exception>
		public async Task<bool> RemoveListener(int listenerID, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(listenerID)
				? await ExecuteAsync(System.Net.Http.HttpMethod.Delete, string.Join(Paths.Separator, Paths.Listeners, listenerID), string.Empty, ct)
				: throw new ArgumentException($"Invalid ListenerID provided");

		/// <summary>
		/// Update the info for a listener
		/// </summary>
		/// <param name="listener"></param>
		/// <param name="ct"></param>
		/// <exception cref="NotImplementedException"></exception>
		public async Task<bool> UpdateListener(Types.Listener listener, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Put, string.Join(Paths.Separator, Paths.Listeners, listener.Id), JSON.ConvertToJSON(listener), ct);
	}
}
